pandas/tests/io/parser/test_parse_dates.py

```diff
@@ -0,0 +1,835 @@
+
+
+def test_date_parser_int_bug(all_parsers):
+    # see gh-3071
+    parser = all_parsers
+    data = ("posix_timestamp,elapsed,sys,user,queries,query_time,rows,"
+            "accountid,userid,contactid,level,silo,method\n"
+            "1343103150,0.062353,0,4,6,0.01690,3,"
+            "12345,1,-1,3,invoice_InvoiceResource,search\n")
+
+    result = parser.read_csv(
+        StringIO(data), index_col=0, parse_dates=[0],
+        date_parser=lambda x: datetime.utcfromtimestamp(int(x)))
+    expected = DataFrame([[0.062353, 0, 4, 6, 0.01690, 3, 12345, 1, -1,
+                           3, "invoice_InvoiceResource", "search"]],
+                         columns=["elapsed", "sys", "user", "queries",
+                                  "query_time", "rows", "accountid",
+                                  "userid", "contactid", "level",
+                                  "silo", "method"],
+                         index=Index([Timestamp("2012-07-24 04:12:30")],
+                                     name="posix_timestamp"))
+    tm.assert_frame_equal(result, expected)
+
+
+def test_nat_parse(all_parsers):
+    # see gh-3062
+    parser = all_parsers
+    df = DataFrame(dict({"A": np.asarray(lrange(10), dtype="float64"),
+                         "B": pd.Timestamp("20010101")}))
+    df.iloc[3:6, :] = np.nan
+
+    with tm.ensure_clean("__nat_parse_.csv") as path:
+        df.to_csv(path)
+
+        result = parser.read_csv(path, index_col=0, parse_dates=["B"])
+        tm.assert_frame_equal(result, df)
+
+
+def test_csv_custom_parser(all_parsers):
+    data = """A,B,C
+20090101,a,1,2
+20090102,b,3,4
+20090103,c,4,5
+"""
+    parser = all_parsers
+    result = parser.read_csv(
+        StringIO(data),
+        date_parser=lambda x: datetime.strptime(x, "%Y%m%d"))
+    expected = parser.read_csv(StringIO(data), parse_dates=True)
+    tm.assert_frame_equal(result, expected)
+
+
+def test_parse_dates_implicit_first_col(all_parsers):
+    data = """A,B,C
+20090101,a,1,2
+20090102,b,3,4
+20090103,c,4,5
+"""
+    parser = all_parsers
+    result = parser.read_csv(StringIO(data), parse_dates=True)
+
+    expected = parser.read_csv(StringIO(data), index_col=0,
+                               parse_dates=True)
+    tm.assert_frame_equal(result, expected)
+
+
+def test_parse_dates_string(all_parsers):
+    data = """date,A,B,C
+20090101,a,1,2
+20090102,b,3,4
+20090103,c,4,5
+"""
+    parser = all_parsers
+    result = parser.read_csv(StringIO(data), index_col="date",
+                             parse_dates=["date"])
+    index = date_range("1/1/2009", periods=3)
+    index.name = "date"
+
+    expected = DataFrame({"A": ["a", "b", "c"], "B": [1, 3, 4],
+                          "C": [2, 4, 5]}, index=index)
+    tm.assert_frame_equal(result, expected)
+
+
+# Bug in https://github.com/dateutil/dateutil/issues/217
+# has been addressed, but we just don't pass in the `yearfirst`
+@pytest.mark.skip(reason="yearfirst is not surfaced in read_*")
```

Comment:
By not surfacing `yearfirst`, we implicitly always have `yearfirst=False` (we want `yearfirst=True` here)

Question is: do we want to surface this parameter to `read_csv` (adds functionality but adds more bloat to API)?  Or, should we go the other way and deprecate `dayfirst` and require users to pass in a parser to the `date_parser` argument (reduces bloat but will be a breaking change for users)?

Reply:
don't skip, rather xfail

Reply:
Done.  I also see that #2586 is already discussing this.